File: apps/Camera-Vitals/Android/app/src/main/java/ai/zetic/demo/cameravitals/util/RollingStats.kt
package ai.zetic.demo.cameravitals.util

/**
 * Median window followed by an EMA — keeps the displayed BPM steady so a single
 * noisy window never flashes a wild number on screen. (Mirrors iOS MedianEMA.)
 */
class MedianEMA(private val size: Int = 5, private val alpha: Double = 0.3) {
    private val window = ArrayDeque<Double>()
    var value: Double? = null
        private set

    fun update(x: Double): Double {
        window.addLast(x)
        if (window.size > size) window.removeFirst()
        val med = window.sorted()[window.size / 2]
        val v = value
        value = if (v != null) alpha * med + (1 - alpha) * v else med
        return value!!
    }

    fun reset() {
        window.clear()
        value = null
    }
}

File: apps/Camera-Vitals/iOS/CameraVitals/Util/RollingStats.swift
import Foundation

/// Median window followed by an EMA — keeps the displayed BPM steady so a single
/// noisy window never flashes a wild number on screen.
struct MedianEMA {
    private var window: [Double] = []
    private let size: Int
    private let alpha: Double
    private(set) var value: Double?

    init(size: Int = 5, alpha: Double = 0.3) {
        self.size = size
        self.alpha = alpha
    }

    @discardableResult
    mutating func update(_ x: Double) -> Double {
        window.append(x)
        if window.count > size { window.removeFirst() }
        let med = window.sorted()[window.count / 2]
        if let v = value {
            value = alpha * med + (1 - alpha) * v
        } else {
            value = med
        }
        return value!
    }

    mutating func reset() {
        window.removeAll()
        value = nil
    }
}
